### PR TITLE
treewide: get rid of `with lib;` and friends

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Description of changes

Tracks

https://github.com/NixOS/nixpkgs/issues/208242

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
I am cogitating if it is better to split this per file.

File: nixos/maintainers/scripts/cloudstack/cloudstack-image.nix
Comment (thiagokokada) [inline nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: Would be nice to have a PR with only those removals, since this one can be merged really fast.

Comment (AndersonTorres) [reply nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: I am not following.
My idea is that each commit should be buildable. If I merely remove the `with lib` and merge, obviously it will generate errors like "mkDefault not defined".

Comment (thiagokokada) [reply nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: What do you mean? That this commit is broken them? Because it is clear that a few files only remove `with lib;` and do nothing else.

Comment (thiagokokada) [reply nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: I am talking about only about the files that remove `with lib;` and do nothing else BTW.

Comment (AndersonTorres) [reply nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: Good idea.

File: nixos/modules/hardware/all-firmware.nix
Comment (SuperSandro2000) [inline nixos/modules/hardware/all-firmware.nix:5]: IMO this is not an improvement over just prefixing things with ``lib.`` or using ``with lib`` more locally.

Comment (AndersonTorres) [reply nixos/modules/hardware/all-firmware.nix:5]: At this first moment I am more interested in make the expressions compile. More cosmetic modifications can be done later.

Comment (SuperSandro2000) [reply nixos/modules/hardware/all-firmware.nix:5]: I am talking about the general idea of your change. If the next person (for example me) is going to revert this again because maintaining a list of used functions is not worth it, then we didn't achieve anything.

File: nixos/maintainers/scripts/cloudstack/cloudstack-image.nix
Comment (SuperSandro2000) [reply nixos/maintainers/scripts/cloudstack/cloudstack-image.nix:5]: Yes, same idea from me. If we can remove a ``with`` anywhere without having regressions we can immediately merge it. This can also be done treewide to make mistakes in commits unlikely.

This change is uncontroversial for me

File: nixos/modules/hardware/gkraken.nix
Comment (SuperSandro2000) [inline nixos/modules/hardware/gkraken.nix:17]: For me the line before the change was so easy to understand that I can parse it at first glimpse without thinking. I can no longer do that with ``pkgs.`` especially if there are multiple.

File: nixos/modules/hardware/gkraken.nix
Comment (SuperSandro2000) [inline nixos/modules/hardware/gkraken.nix:7]: Each of those functions is used once. The file is not even 20 lines long. Both the inherit and with should be dropped and the three occurrences of lib should just be written out.

File: nixos/modules/testing/service-runner.nix
Comment (SuperSandro2000) [inline nixos/modules/testing/service-runner.nix:125]: This usage of lib is totally fine for me and I see a lot of people doing it when writing modules. IMO we should keep it.

File: nixos/modules/testing/service-runner.nix
Comment (SuperSandro2000) [inline nixos/modules/testing/service-runner.nix:115]: We should not change formatting that changes the code to avoid unnecessary possibilities  regressions.

File: nixos/tests/yabar.nix
Comment (SuperSandro2000) [inline nixos/tests/yabar.nix:5]:
```suggestion
  meta.maintainers = with lib.maintainers; [ ];
```

Yes, this is technically unused but it is often kept to show that there is no maintainer. It would be fine for me to remove those to make it easier for tooling to handle this.

Also pkgs.lib should not be used when we can easily get a lib input.

File: nixos/tests/uptime-kuma.nix
Comment (SuperSandro2000) [inline nixos/tests/uptime-kuma.nix:5]:
```suggestion
  meta.maintainers = with lib.maintainers; [ julienmalka ];
```
When there are multiple maintainers we would end up with code duplication.
I don't see a reason why a meta.maintainers with only one entry should be differently formatted then one with multiple, hence I am suggesting to make it ready for multiple maintainers and don't suggest code duplication.

File: nixos/tests/tor.nix
Comment (SuperSandro2000) [inline nixos/tests/tor.nix:1]: The placement of this with lib is unexpected. If it wasn't removed I would at least have moved it to a more normal position.

File: nixos/tests/redis.nix
Comment (SuperSandro2000) [inline nixos/tests/redis.nix:10]: I suggest to either prefix the functions with ``lib.`` or move the with lib to the function.

File: nixos/tests/n8n.nix
Comment (SuperSandro2000) [inline nixos/tests/n8n.nix:10]: IMO this code is now actually worse and not understandable for a beginner. I am against changing multi maintainer entries to use a function instead of being a list which uses with very locally.

File: nixos/tests/gnome-xorg.nix
Comment (SuperSandro2000) [inline nixos/tests/gnome-xorg.nix:3]: That's a good change 👍🏼

File: nixos/tests/gitlab.nix
Comment (SuperSandro2000) [inline nixos/tests/gitlab.nix:29]: I would have done the same

File: nixos/tests/fluidd.nix
Comment (SuperSandro2000) [inline nixos/tests/fluidd.nix:5]: Great change 👍🏼

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
We can't merge this before we don't have some consensus otherwise the next person is going to revert it and then we all are not making good use of our time.

I would suggest that we split out all the uncontroversial changes into separate, treewide PRs so that they can be merged faster and the merge conflicts are reduced. 

Also please do not self merge this PR.

File: nixos/modules/hardware/all-firmware.nix
Comment (AndersonTorres) [reply nixos/modules/hardware/all-firmware.nix:5]: The "idea" of my change is just getting rid of `with`. 

For this draft phase, I believe it is easier to use `inherit (lib) every thing that was called;`. It will help a future automation.

I have no problem in prefixing `lib.` This can be done with the help of an Emacs quick&dirty magic keystroke. Indeed I will do this tonight.

File: nixos/modules/hardware/gkraken.nix
Comment (AndersonTorres) [reply nixos/modules/hardware/gkraken.nix:7]: Being redundant, my primary concern here is to make the code compile without `with`.

After this, it will be easy to prefix lib.

OTOH I don't believe this is bad code. It is similar to other programming languages like C++ namespaces and Python lib imports.

File: nixos/tests/n8n.nix
Comment (AndersonTorres) [reply nixos/tests/n8n.nix:10]: Beginners mostly copy and paste examples. Many of them still tops `with lib;` without knowing why - ans this is demonstrated by the use of `lib.mdDoc`.

**Comment by AndersonTorres** [comment]
Small steps.

File: nixos/modules/hardware/all-firmware.nix
Comment (SuperSandro2000) [reply nixos/modules/hardware/all-firmware.nix]: > And it is depressing to look at `meta = with pkgs.lib.maintainers; { maintainers = []; };`. Such a heavy machinery to empty lists?

It is empty to convey that the package needs a maintainer.

Comment (AndersonTorres) [reply nixos/modules/hardware/all-firmware.nix]: > It is empty to convey that the package needs a maintainer.

What is the difference for `meta.maintainers = [ ];`?

**Comment by AndersonTorres** [comment]
Closing because headache.

**Pull request closed without merging**
